File: sim.py
from src.main import calc_income, calc_employment
# from PyQt5 import QtWidgets
from src.building import *
from src.constants import *

# class Day(QtWidgets.QWidget):
#     def __init__(self, parent=None):
#         super().__init__(parent)
#         self.layout = QtWidgets.QGridLayout(self)
#         self.l_bal = QtWidgets.QLabel(self)
#         self.l_employ = QtWidgets.QLabel(self)
#         self.l_income = QtWidget.QLabel(self)
#         self.
#         self.setLayout(self.layout)

# class Main(QtWidgets.QWidget):
#     def __init__(self):
#         super().__init__()
#         self.layout = QtWidgets.QHBoxLayout(self)
#         self.layout.addWidget(Day(self))
#         self.setLayout(self.layout)
#         self.show()

# if __name__ == "__main__":
#     app = QtWidgets.QApplication([])
#     ex = Main()
#     app.exec_()

import datetime
d = datetime.date.today()

class Economy:
    def __init__(self, buildings=None, balance=40000, loans=None):
        if buildings is None:
            self.buildings = []
        else:
            self.buildings = buildings
        if loans is None:
            self.loans = []
        else:
            self.loans = loans

        self.bal = balance

    def add_building(self, building):
        self.buildings.append(building)
        self.bal -= building[0].cost() * building[1]

    def add_building2(self, btype, count=1, size=None):
        self.add_building((Building(btype, d, self.lorentz(), size), count))

    def add_loan(self, loan):
        self.loans.append(loan)
        self.bal += loan[0]

    def get_loans(self):
        return sum([n[0] for n in self.loans])

    def pay_for_loan(self, idx, amt):
        self.loans[idx][0] -= amt
        self.bal -= amt

    def income(self):
        return calc_income(self.get_data())[0]

    def employ(self):
        return calc_employment(self.get_data())
        
    def get_data(self):
        buildings = []
        for b in self.buildings:
            for i in range(b[1]):
                buildings.append(b[0])
        
        return {"regions": {"a": {"buildings": buildings}}}

    def day(self):
        income = self.income()
        employment = self.employ()
        oldbal = self.bal
        print(f"Bal: {self.bal:.2f}, Emp: {employment:.4f}, Inc: {income:.2f}, Exc: {oldbal + income:.2f}")
        self.bal += income
        if self.bal < 0:
            self.bal += self.bal * OVERDRAFT_INTEREST

        for l in self.loans:
            l[0] *= 1 + l[1]

        # print(f"Bal: {self.bal:.2f}, Emp: {employment:.2f}, Inc: {income:.2f}, Exc: {oldbal + income:.2f}")
        # print("Loans:")
        # for i, l in enumerate(self.loans):
            # print(f"{i}: {l[0]:.2f} @ {l[1] * 100:.2f}%")

        # print("\n")
    def lorentz(self):
        return Building.get_lorentz(self.income())

"""
things_to_build = [(Building(FARMING, d), 50)] * 32 + [
    (Building(FARMING, d), 19),
    (Building(HOUSE, d, 4), 1),
    (Building(HOUSE, d, 4), 1),
    (Building(HOUSE, d, 4), 1),
    (Building(HOUSE, d, 4), 1),
    (Building(HOUSE, d, 2), 1),
    (Building(HOUSE, d, 2), 1),
    (Building(HOUSE, d, 2), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
    (Building(OFFICE, d), 1),
]

def do_something(eco, val):
    pass

def fitness(val):
    eco = Economy(buildings=[
            (Building(HOUSE, d, 4), 7),
            (Building(AIRPORT, d, 100), 1),
            (Building(MARKET_STALL, d), 1),
            (Building(AIRPORT, d, 109), 1),
            (Building(POLICE_STATION, d), 1),
            (Building(HOSPITAL, d), 1),
            (Building(POST_OFFICE, d), 1),
            (Building(OFFICE, d), 9),
            (Building(FARMING, d), 300),
        ], balance=-4118.68,
           loans=[[13147.8, 0.042]])

    for i in range(10):
        eco.day()
        do_something(eco, val)

    return eco.income() if eco.bal > 0 else 0
print(len(things_to_build))
print(fitness(0))
"""

eco = Economy(buildings=[
    (Building(BType.AIRPORT, d, 1, 100), 1),
    (Building(BType.AIRPORT, d, 1, 109), 1),
    (Building(BType.AIRPORT, d, 1, 90), 1),
    (Building(BType.HOUSE, d, 1, 4), 10),
    (Building(BType.HOUSE, d, 1, 2), 2),
    (Building(BType.MARKET_STALL, d, 1), 1),
    (Building(BType.POLICE_STATION, d, 1), 1),
    (Building(BType.HOSPITAL, d, 1), 1),
    (Building(BType.POST_OFFICE, d, 1), 1),
    (Building(BType.OFFICE, d, 1), 13),
    (Building(BType.FARMING, d, 1), 1919),
], balance=-3510.0)

# 8100.817722685765 1454.5103758301166
def good_one():
    eco.add_building((Building(BType.HOUSE, d, eco.lorentz(), 2), 1))
    # print(f"Bal: {eco.bal:.2f}, Emp: {eco.employ():.2f}, Inc: {eco.income():.2f}")
    eco.day()
    eco.add_building2(BType.HOUSE, 1, 4)
    eco.day()
    eco.add_building2(BType.HOUSE, 1, 1)
    eco.add_building2(BType.FARMING, 357)
    eco.day() # 3
    eco.add_loan([1000, 0.02])
    eco.add_building2(BType.POLICE_STATION, 1)
    eco.add_building2(BType.HOUSE, 1, 2)
    eco.day()
    eco.add_building2(BType.HOUSE, 1, 2)
    eco.add_building2(BType.OFFICE, 1)
    eco.pay_for_loan(0, 1000*1.02)
    eco.day()

def weird_one():
    eco.add_loan([3000, 0.02])
    eco.add_building((Building(BType.HOUSE, d, eco.lorentz(), 4), 1))
    # print(f"Bal: {eco.bal:.2f}, Emp: {eco.employ():.2f}, Inc: {eco.income():.2f}")
    eco.day()
    eco.add_building2(BType.HOUSE, 1, 2)
    eco.pay_for_loan(0, 3000*1.02)
    eco.day()
    eco.add_building2(BType.HOUSE, 1, 1)
    eco.add_building2(BType.FARMING, 357)
    eco.day() # 3
    eco.add_loan([1000, 0.02])
    eco.add_building2(BType.POLICE_STATION, 1)
    eco.add_building2(BType.HOUSE, 1, 2)
    eco.day()
    eco.add_building2(BType.HOUSE, 1, 2)
    eco.add_building2(BType.OFFICE, 1)
    eco.pay_for_loan(0, 1000*1.02)
    eco.day()

weird_one()
print(eco.income(), eco.bal - eco.get_loans())

File: sim.rs
#[derive(PartialEq, Clone, Copy)]
enum BType {
    RailwayStation,
    MarketStall,
    PoliceStation,
    PostOffice,
    SmallStore,
    Hospital,
    FireStation,
    SuperStore,
    Pier,
    Dock,
    Quarry,
    SmallFactory,
    LargeFactory,
    Farming,
    NavalDockyard,
    Mills,
    Airbase,
    SupplyHub,
    Reactor,
    ElectricalGeneration,
    Airport,
    House,
    Office,
    MetroStation,
}

struct BuildingInfo {
    wage: f64,
    employees: f64,
    cost: f64,
}

#[derive(Clone, Copy)]
struct Building {
    ty: BType,
    size: u32,
    count: f64,
}

struct Loan {
    amt: f64,
    interest: f64,
}

struct Economy {
    buildings: Vec<Building>,
    loans: Vec<Loan>,
    bal: f64,
}

const INFO: [BuildingInfo; 24] = [
    BuildingInfo { wage: 13.5, employees: 2.0,    cost: 2808.0 },
    BuildingInfo { wage: 11.0, employees: 1.0,    cost: 1144.00 },
    BuildingInfo { wage: 19.6, employees: 3.0,    cost: 6115.20 },
    BuildingInfo { wage: 12.5, employees: 2.0,    cost: 2600.00 },
    BuildingInfo { wage: 11.5, employees: 2.0,    cost: 2392.00 },
    BuildingInfo { wage: 26.0, employees: 3.0,    cost: 8112.00 },
    BuildingInfo { wage: 19.0, employees: 2.0,    cost: 3952.00 },
    BuildingInfo { wage: 12.2, employees: 4.0,    cost: 5075.20 },
    BuildingInfo { wage: 14.0, employees: 1.0,    cost: 1456.00 },
    BuildingInfo { wage: 15.8, employees: 2.0,    cost: 3286.40 },
    BuildingInfo { wage: 11.2, employees: 2.0,    cost: 2329.60 },
    BuildingInfo { wage: 14.5, employees: 6.0,    cost: 9048.00 },
    BuildingInfo { wage: 15.5, employees: 12.0,   cost: 19344.0 },
    BuildingInfo { wage: 18.5, employees: 1.0/162.0,cost: 11.9    },
    BuildingInfo { wage: 15.5, employees: 1.0,    cost: 1612.00 },
    BuildingInfo { wage: 0.0,  employees: 12.0,   cost: 25500.0 },
    BuildingInfo { wage: 17.5, employees: 4.0,    cost: 7280.00 },
    BuildingInfo { wage: 10.5, employees: 1.0,    cost: 1092.00 },
    BuildingInfo { wage: 20.5, employees: 3.0,    cost: 6396.00 },
    BuildingInfo { wage: 12.5, employees: 2.0,    cost: 2600.00 },
    BuildingInfo { wage: 18.0, employees: 0.0,    cost: 69.0    },
    BuildingInfo { wage: 0.0,  employees: 0.0,    cost: 69420.0 },
    BuildingInfo { wage: 18.0, employees: 1.0,    cost: 1872.00 },
    BuildingInfo { wage: 12.9, employees: 1.0,    cost: 1341.60 }
];
const ROI: f64 = 13.0;
const OVERDRAFT_INTEREST: f64 = 0.30;

impl Building {
    fn new_sized(ty: BType, size: u32, count: u32) -> Self {
        Self {
            ty, size, count: count as f64
        }
    }
    
    fn new(ty: BType, count: u32) -> Self {
        Self {
            ty, size: 0, count: count as f64
        }
    }

    fn cost(&self) -> f64 {
        if self.ty == BType::Airport {
            ROI * self.income()
        } else if self.ty == BType::House {
            (match self.size {
                1 => 1500.0,
                2 => 3000.0,
                4 => 6000.0,
                6 => 9000.0,
                _ => unreachable!()
            }) * self.count
        } else {
            INFO[self.ty as usize].cost * self.count
        }
    }

    fn wage(&self) -> f64 {
        INFO[self.ty as usize].wage
    }

    fn employees(&self) -> f64 {
        if self.ty == BType::Airport {
            (self.size as f64) / 20.0 * self.count
        } else {
            INFO[self.ty as usize].employees * self.count
        }
    }

    fn income(&self) -> f64 {
        self.wage() * self.employees() * 8.0
    }
}

impl Economy {
    fn new(buildings: Vec<Building>, bal: f64, loans: Vec<Loan>) -> Self {
        Self {
            buildings, bal, loans
        }
    }

    fn add_building(&mut self, b: Building) {
        self.bal -= b.cost();
        self.buildings.push(b);
    }
    fn add_loan(&mut self, loan: Loan) {
        self.bal += loan.amt;
        self.loans.push(loan);
    }
    fn pay_for_loan(&mut self, idx: usize, amt: f64) {
        self.loans[idx].amt -= amt;
        self.bal -= amt;
    }

    fn employ(&self) -> f64 {
        let mut people = 0.0;
        let mut jobs = 0.0;
        for b in self.buildings.iter() {
            if b.ty == BType::House {
                people += b.size as f64 * b.count;
            } else {
                jobs += b.employees();
            }
        }
        jobs / people
    }
    
    fn income(&self) -> f64 {
        let employment = self.employ();
        let gross: _ = self.buildings.iter().map(|b| b.income()).sum::<f64>();
        if employment > 1.0 {
            gross / employment
        } else {
            gross * employment
        }
    }

    fn day(&mut self) {
        let inc = self.income();
        self.bal += inc;
        if self.bal < 0.0 {
            self.bal += self.bal * OVERDRAFT_INTEREST;
        }

        for l in self.loans.iter_mut() {
            l.amt *= 1.0 + l.interest;
        }
        // println!("Bal: {}, Emp: {}, Inc: {}", self.bal, self.employ(), inc);
    }
}

const N: usize = 10;

fn fitness(to_do: [Vec<Building>; N]) -> f64 {
    let mut eco = Economy::new(
        vec![
            Building::new_sized(BType::House, 4, 7),
            Building::new_sized(BType::Airport, 100, 1),
            Building::new_sized(BType::Airport, 109, 1),
            Building::new(BType::MarketStall, 1),
            Building::new(BType::PoliceStation, 1),
            Building::new(BType::Hospital, 1),
            Building::new(BType::PostOffice, 1),
            Building::new(BType::Office, 9),
            Building::new(BType::Farming, 300),
        ],
        -4118.68,
        vec![
            Loan { amt: 13147.8, interest: 0.042 }
        ]
    );
    eco.day();
    for i in 0..N {
        for b in to_do[i].iter() {
            eco.add_building(*b);
        }
        eco.day();
    }
    eco.income()
}

fn main() {
    
}
